src/adaptors/mod.rs

```diff
@@ -812,6 +812,72 @@ impl_tuple_combination!(Tuple10Combination Tuple9Combination; a b c d e f g h i)
 impl_tuple_combination!(Tuple11Combination Tuple10Combination; a b c d e f g h i j);
 impl_tuple_combination!(Tuple12Combination Tuple11Combination; a b c d e f g h i j k);
 
+/// An iterator to iterate through all combinations in an iterator that produces arrays
+/// of a specific size.
+///
+/// See [`.array_combinations()`](crate::Itertools::array_combinations) for more
+/// information.
+#[derive(Debug, Clone)]
+pub struct ArrayCombinations<I: Iterator, const R: usize> {
+    iter: I,
+    buf: Vec<I::Item>,
```

Comment:
Wouldn't it be better to use an [`ArrayVec`](https://docs.rs/arrayvec/0.7.1/arrayvec/struct.ArrayVec.html) here, and so avoid heap allocations entirely?

Reply:
IIRC, `buf` caches the elements from `iter`, so I think it is impossible to impose an upper bound on the length here.

Reply:
Ah yes, quite right.

Actually, my current use-case is to generate pairs of items from an iterator over a large array; so buffering all items is quite costly... in this case, my iterator implements `Clone` so I can instead hold `R` clones of the iterator and entirely avoiding dynamic/heap allocations:

```rust
struct ArrayCombinations<I: Iterator, const R: usize> {
    /// Each element of this array is an iterator that can be inspected for
    /// the corresponding element of the resulting combo.
    ///
    /// Peeking each iterator yields the last emitted value, such that
    /// invoking `next()` first advances the state before emitting the
    /// result.  For example, given `[1,2,3,4,5].array_combinations::<3>()`,
    /// the initial state of `iters` is such that peeking each iterator element
    /// will yield `[1, 2, 2]` (becoming `[1, 2, 3]` after `next()` is called).
    iters: [Peekable<I>; R],
}

impl<I, const R: usize> ArrayCombinations<I, R>
where
    I: Iterator + Clone,
    I::Item: Clone,
{
    pub fn new(iter: I) -> Self {
        let mut iters = ArrayVec::new();
        iters.push(iter.peekable());
        for idx in 1..R {
            let last = iters.last_mut().unwrap();
            if idx > 1 {
                last.next();
            }
            let last = last.clone();
            iters.push(last);
        }
        Self { iters: iters.into_inner().ok().unwrap() }
    }
}

impl<I, const R: usize> Iterator for ArrayCombinations<I, R>
where
    I: Iterator + Clone,
    I::Item: Clone,
{
    type Item = [I::Item; R];

    fn next(&mut self) -> Option<Self::Item> {
        'search: for idx in (0..R).rev() {
            self.iters[idx].next();
            if self.iters[idx].peek().is_some() {
                let mut clone = self.iters[idx].clone();
                for reset in self.iters[idx+1..].iter_mut() {
                    clone.next();
                    match clone.peek() {
                        Some(_) => *reset = clone.clone(),
                        None => continue 'search,
                    }
                }
                break;
            }
        }

        self.iters
            .iter_mut()
            .map(Peekable::peek)
            .map(Option::<&_>::cloned)
            .collect::<Option<ArrayVec<_, R>>>()
            .map(ArrayVec::into_inner)
            .map(Result::ok)
            .flatten()
    }
}
```